### PR TITLE
Enable Management of NCS Pathways

### PR DESCRIPTION
**Pull request by gkahiu**
Enable adding, editing or removing NCS pathways.

The behavior of editing NCS pathways has changed and it will only be applied under the following conditions:

- The NCS pathway is a default one i.e. `ncs.user_defined = True`. These are the ones defined under `definitions/defaults.py`
- The main layer path and carbon paths could not be found hence, the module will try to default to the paths defined under `ncs_pathways` and `ncs_carbon` respectively

Incorporates features outlined in issue #106.

### CODE REVIEW

File: src/cplus_plugin/conf.py [hunk 1]
```diff
@@ -715,20 +736,30 @@ def update_ncs_pathway(self, ncs_pathway: NcsPathway):
...
+                abs_path = str(os.path.normpath(abs_path))
+                ncs_pathway.path = abs_path
+
+            # Carbon location
+            abs_carbon_paths = []
+            for cb_path in ncs_pathway.carbon_paths:
+                cp = Path(cb_path)
+                # Similarly, if the given carbon path does not exist then try
+                # to use the default one in the ncs_carbon directory.
+                if not cp.exists():
```
Comment (Samweli) [inline src/cplus_plugin/conf.py:755]: Same as above

File: src/cplus_plugin/models/helpers.py [hunk 2]
```diff
@@ -141,10 +141,18 @@ def create_ncs_pathway(source_dict) -> typing.Union[NcsPathway, None]:
     :rtype: NcsPathway
     """
     ncs = create_layer_component(source_dict, NcsPathway)
+
+    # We are checking because of the various iterations of the attributes
+    # in the NcsPathway class where some of these attributes might
+    # be missing.
     carbon_paths_attr = "carbon_paths"
```
Comment (Samweli) [inline src/cplus_plugin/models/helpers.py:148]: @gkahiu now that we have time for code structure improvements, lets make minor/small improvements in the codebase. We need to define all the strings literal in the `constants.py` file eg. for here the `carbon_paths` string should have its own constant variable.

Comment (Samweli) [inline src/cplus_plugin/models/helpers.py:148]:
```suggestion
    carbon_paths_attr = constants.CARBON_PATHS
```

File: src/cplus_plugin/main.py [hunk 3]
```diff
@@ -359,6 +359,9 @@ def initialize_model_settings():
         # Create priority weighting layers subdirectory
         FileUtils.create_pwls_dir(base_dir)
 
+    # Use carbon coefficient values for older-version NCS pathways
+    carbon_coefficient = settings_manager.get_value(Settings.CARBON_COEFFICIENT, 0)
```
Comment (Samweli) [inline src/cplus_plugin/main.py:363]:
```suggestion
    carbon_coefficient = settings_manager.get_value(Settings.CARBON_COEFFICIENT, 0.0)
```

**Review by Samweli — COMMENTED** [review]
@gkahiu can you also include the issue number that this PR is fixing in the PR comment, thank you!

File: src/cplus_plugin/conf.py [hunk 4]
```diff
@@ -715,20 +736,30 @@ def update_ncs_pathway(self, ncs_pathway: NcsPathway):
...
-            abs_carbon_path = f"{base_dir}/{NCS_CARBON_SEGMENT}/" f"{cp.name}"
-            abs_carbon_path = str(os.path.normpath(abs_carbon_path))
-            abs_carbon_paths.append(abs_carbon_path)
-        ncs_pathway.carbon_paths = abs_carbon_paths
+        # Pathway location for default pathway
+        if not ncs_pathway.user_defined:
+            p = Path(ncs_pathway.path)
+            # Only update if path does not exist otherwise
+            # fallback to check under base directory.
+            if not p.exists():
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/conf.py:750]: Thanks @Samweli for the feedback. If we include the additional checks then we would exclude cases where the base directory has changed. Am assuming the user will have the option of changing the location of map layers for default pathways to custom locations; this check is just a fallback mechanism to use the default datasets (assuming these exist in the base directory) if the current layer path does not exist. But am happy to discuss further.

File: src/cplus_plugin/conf.py (hunk 1 above)
Comment (gkahiu) [reply src/cplus_plugin/conf.py:761]: Same as above but we can discuss.

File: test/model_data_for_testing.py [hunk 5]
```diff
@@ -127,4 +128,5 @@ def get_test_scenario_result() -> ScenarioResult:
     "layer_type": 0,
     "user_defined": True,
     "carbon_paths": [],
+    "carbon_coefficient": 0.00,
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply test/model_data_for_testing.py]: Done.

File: src/cplus_plugin/models/helpers.py (hunk 2 above)
Comment (gkahiu) [reply src/cplus_plugin/models/helpers.py]: Done.

Comment (gkahiu) [reply src/cplus_plugin/models/helpers.py]: Done.

File: src/cplus_plugin/models/helpers.py [hunk 6]
```diff
@@ -141,10 +141,18 @@ def create_ncs_pathway(source_dict) -> typing.Union[NcsPathway, None]:
...
     if carbon_paths_attr in source_dict:
         ncs.carbon_paths = source_dict["carbon_paths"]
 
+    carbon_coefficient_attr = "carbon_coefficient"
```
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/models/helpers.py]: Done

File: src/cplus_plugin/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gkahiu) [reply src/cplus_plugin/main.py]: Done.

Comment (gkahiu) [reply src/cplus_plugin/main.py]: Done.

**Comment by Samweli** [comment]
Thanks @gkahiu

**Pull request merged**